farm-bot/codebase/contracts/farm-bot.sol

```diff
@@ -2,62 +2,205 @@
 pragma solidity >=0.5.0 <0.9.0;
 
 import "hardhat/console.sol";
-import "./ubeswap-farming/contracts/StakingRewards.sol"; // does the ierc20 import too
 
-contract FarmBot {
-    mapping(address => uint256) public fpBalances; // maps address to "Farm Point" balances. todo eventually make private
-    // FP balances ONLY changes on deposits/withdrawals, NOT when FarmBot claims rewards. This helps save gas.
-    // NOTE: the thing that DOES change is the VALUE of an FP relative to an LP token
+import "./ubeswap-farming/contracts/StakingRewards.sol";
+import "./ubeswap/contracts/uniswapv2/interfaces/IUniswapV2Router02.sol";
+import "./ubeswap/contracts/uniswapv2/interfaces/IUniswapV2Pair.sol";
+import "./ubeswap-farming/contracts/Owned.sol";
+import "./FarmbotERC20.sol";
 
+contract FarmBot is Owned, FarmbotERC20 {
     uint256 public lpTotalBalance; // total number of LP tokens owned by Farm Bot
-    uint256 public fpTotalBalance; // total number of "Farm Points" issued by Farm Bot
 
-    address public tokenAddress = 0xf3098223845F37Ffe4B3A066F2D38A0587317269; // mcUSD-Celo
-    StakingRewards public farm = StakingRewards(0x299f31f48D4667a6f68E4331dB05212d57Cc7f80); // alfajores farm for mcUSD-Celo
+    StakingRewards public stakingRewards;
+
+    IERC20 public rewardsToken;
+
+    IUniswapV2Pair public stakingToken; // LP that's being staked
+    IERC20 public stakingToken0; // LP token0
+    IERC20 public stakingToken1; // LP token1
+
+    IUniswapV2Router02 public router; // Router address
+
+    // Paths for swapping; can be updated by owner
+    address[] public path0; // Path to use when swapping rewardsToken to token0. If len < 2, we assume rewardsToken == token0
+    address[] public path1; // Path to use when swapping rewardsToken to token1. If len < 2, we assume rewardsToken == token1
+
+    // Acceptable slippage when swapping/minting LP; can be updated by owner
+    uint256 public slippageNumerator = 99;
+    uint256 public slippageDenominator = 100;
```

Comment:
I think `slippageNumerator` and `slippageDenominator` are always used together. Why not just use one field, `slippageFraction`?

Reply:
Not sure how that would work when all our types are ints. We have to multiply and divide by the numerator/denominator in order to get a "fraction".. Unless I'm misunderstanding something.

Reply:
whoops, forgot this about solidity